src/lib.rs

```diff
@@ -30,70 +33,55 @@ mod defmt {
     pub use super::{error, info};
 }
 
+/// Alias to the Pin tuple used in `Spi<…>`
+pub type Pins<P, MISO, MOSI, SCLK> = (
+    Pin<<MISO as AnyPin>::Id, <P as PIOExt>::PinFunction, <MISO as AnyPin>::Pull>,
+    Pin<<MOSI as AnyPin>::Id, <P as PIOExt>::PinFunction, <MOSI as AnyPin>::Pull>,
+    Pin<<SCLK as AnyPin>::Id, <P as PIOExt>::PinFunction, <SCLK as AnyPin>::Pull>,
+);
+
+/// PIO based Spi driver.
 pub struct Spi<'pio, P, SMI, MISO, MOSI, SCLK, const DS: u8>
 where
-    P: PIOExt + FunctionConfig,
+    P: PIOExt,
     SMI: StateMachineIndex,
-    MISO: PinId,
-    MOSI: PinId,
-    SCLK: PinId,
-    Function<P>: ValidPinMode<MISO> + ValidPinMode<MOSI> + ValidPinMode<SCLK>,
+    MISO: AnyPin,
```

Comment:
This driver only ever uses the Pin when it's in Pio mode, should we store the pins as this type instead or does that not provide any benefit here?
The `::new()` fn can accept AnyPin and convert it to the internal type if you don't want to expose the user to this complexity.

Reply:
We do store them as `Pin<_, P::PinFunction, _>` via the `Pins<…>` type alias.
New needs to them to be `FuntionNull` at start to avoid (or at least limit) glitches on bring up.

So the `::new()` function accepts `AnyPin<Function = FunctionNull>` (that is what is captured in the `Spi<…>` type) and converts it internally to what is actually used.

I tried avoiding capturing the `FunctionNull` state but in all my attempts the compiler lost the ability to infer the correct typets one way or another.